mkosi/run.py

```diff
@@ -219,44 +264,27 @@ def spawn(
     if pass_fds:
         env["LISTEN_FDS"] = str(len(pass_fds))
 
-    def preexec() -> None:
-        if foreground:
-            make_foreground_process()
-        if preexec_fn:
-            preexec_fn()
-
-        if not pass_fds:
-            return
+    with sandbox as sbx:
+        prefix = [os.fspath(x) for x in sbx]
 
-        # The systemd socket activation interface requires any passed file descriptors to start at '3' and
-        # incrementally increase from there. The file descriptors we got from the caller might be arbitrary,
-        # so we need to move them around to make sure they start at '3' and incrementally increase.
-        for i, fd in enumerate(pass_fds):
-            # Don't do anything if the file descriptor is already what we need it to be.
-            if fd == SD_LISTEN_FDS_START + i:
-                continue
-
-            # Close any existing file descriptor that occupies the id that we want to move to. This is safe
-            # because using pass_fds implies using close_fds as well, except that file descriptors are closed
-            # by python after running the preexec function, so we have to close a few of those manually here
-            # to make room if needed.
-            try:
-                os.close(SD_LISTEN_FDS_START + i)
-            except OSError as e:
-                if e.errno != errno.EBADF:
-                    raise
+        if prefix:
+            prfd, pwfd = os.pipe2(os.O_CLOEXEC)
 
-            nfd = fcntl.fcntl(fd, fcntl.F_DUPFD, SD_LISTEN_FDS_START + i)
-            # fcntl.F_DUPFD uses the closest available file descriptor ID, so make sure it actually picked
-            # the ID we expect it to pick.
-            assert nfd == SD_LISTEN_FDS_START + i
+            # Make sure the write end of the pipe (which we pass to the subprocess) is higher than all the
+            # file descriptors we'll pass to the subprocess, so that it doesn't accidentally get closed by
+            # the logic in preexec().
+            if pass_fds:
+                pwfd = fd_move_above(pwfd, list(pass_fds)[-1])
 
-    with sandbox as sbx:
-        prefix = [os.fspath(x) for x in sbx]
+            exec_prefix = ["--exec-fd", str(SD_LISTEN_FDS_START + len(pass_fds)), "--"]
```

Comment:
I think this would be slightly more readable.
```suggestion
            exec_prefix = ["--exec-fd", f"{SD_LISTEN_FDS_START + len(pass_fds)}", "--"]
```
Looking at this I had to remember that `SD_LISTEN_FDS_START` is 3 to remember that this expression makes sense, because the parentheses are a bit non-obvious. The f-string is a bit unnecessary, but sidesteps this.